sefaria.js

```diff
@@ -220,7 +259,130 @@ Sefaria = {
+      Sefaria.cacheCurrVersionsBySection(versions, ref);
+    }
+    for (let [lang, vtitle] of Object.entries(versions)) {
+      const versionText = await Sefaria.loadOfflineSectionByVersionWithCacheAndFallback(fileNameStem, lang, vtitle, defaultVersions[lang]);
+      // versionText may be depth-3. extract depth-2 if necessary.
+      textByLang[lang] = Sefaria.getSectionFromJsonData(ref, versionText);
+    }
+
+    const fullSection = {...metadata};
+    delete fullSection.links;
+    fullSection.content = [];
+    const sectionLen = Math.max(...Object.values(textByLang).map(x => x.length))
+    for (let i = 0; i < sectionLen; i++) {
+      fullSection.content.push({
+        segmentNumber: i+1+"",
+        links: metadata.links?.[i] || [],
+        text: textByLang?.en?.[i] || "",
+        he: textByLang?.he?.[i] || "",
+      });
+    }
+
+    Sefaria._jsonSectionData[key] = fullSection;
+    return fullSection;
+  },
+  loadOfflineSectionByVersionWithCacheAndFallback: async function(fileNameStem, lang, vtitle, defaultVTitle) {
+    /**
+     * tries to load `vtitle`. If it fails, falls back on default and if that fails, throws an error that this version
+     * isn't offline
+     * if defaultVTitle is falsy, only try to load `vtitle` and if that fails throw error
+     */
+    try {
+      return await Sefaria.loadOfflineSectionByVersionWithCache(fileNameStem, lang, vtitle);
+    } catch(error) {
+      if (!defaultVTitle) {
+        throw ERRORS.NOT_OFFLINE;
+      }
+      try {
+        return await Sefaria.loadOfflineSectionByVersionWithCache(fileNameStem, lang, defaultVTitle);
+      } catch(error) {
+        throw ERRORS.NOT_OFFLINE;
+      }
+    }
+  },
+  loadOfflineSectionByVersionWithCache: async function(fileNameStem, lang, vtitle) {
+    const key = `${fileNameStem}|${lang}|${vtitle}`;
+    const cached = Sefaria._jsonSectionData[key];
+    if (cached) { return cached; }
+    const text = await Sefaria.loadOfflineSectionByVersion(fileNameStem, lang, vtitle);
+    Sefaria._jsonSectionData[key] = text;
+    return text;
+  },
+  loadOfflineSectionByVersion: async function(fileNameStem, lang, vtitle) {
+    /**
+     * Assumption is zip file was already unzipped in loading of metadata
+     * We also already know the fileNameStem from loading of metadata
+     */
+    const jsonPath = Sefaria._JSONSectionPath(fileNameStem, vtitle, lang);
+    return await Sefaria._loadJSON(jsonPath);
+  },
+  loadOfflineSectionMetadataWithCache: async function(ref) {
+    const key = `${ref}|metadata`;
+    const cached = Sefaria._jsonSectionData[key];
+    if (cached) { return cached; }
+    let metadata;
+    try {
+      metadata = await Sefaria.loadOfflineSectionMetadata(ref);
+    } catch(error) {
+      throw ERRORS.OFFLINE_LIBRARY_NOT_COMPATIBLE_WITH_V7;
+    }
+    Sefaria._jsonSectionData[key] = metadata;
+    return metadata;
+  },
+  loadOfflineSectionMetadata: async function(ref) {
+    if (Sefaria.shouldLoadFromApi()) {
```

Comment:
Is the awareness of online features such as the api perhaps to high level for a lower level functions like this?

Reply:
Also, I think the dealing with file system level concerns such as unzipping should be done separately and this file left only to handle an better defined and existing metadata file per section.